### PR TITLE
Add unique predicates

### PR DESCRIPTION
**Pull request by francescoo22**
This PR just adds the unique predicates. These predicates will be added in pre/post-conditions in an upcoming PR.

### CODE REVIEW

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/FieldEmbedding.kt [hunk 1]
```diff
@@ -45,23 +49,29 @@ interface FieldEmbedding {
...
 class UserFieldEmbedding(
     override val name: ScopedKotlinName,
     override val type: TypeEmbedding,
     override val symbol: FirPropertySymbol,
+    override val isUnique: Boolean,
 ) : FieldEmbedding {
     override val viperType = Type.Ref
     override val accessPolicy: AccessPolicy = if (symbol.isVal) AccessPolicy.ALWAYS_READABLE else AccessPolicy.ALWAYS_INHALE_EXHALE
+    override val includeInUniquePredicate: Boolean
+        get() = isUnique || type.isPrimitive
```
Comment (jesyspa) [inline plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/FieldEmbedding.kt:64]: I don't think this is quite right.  A unique predicate should give one permission to read and modify the members.  That includes primitives, but it also includes references.  The only special treatment unique members get is that you also get access *their* members as well, via a nested unique predicate.  For non-unique members, you get their RO predicate.

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/FieldEmbedding.kt [hunk 2]
```diff
@@ -45,23 +49,29 @@ interface FieldEmbedding {
...
     override val unfoldToAccess: Boolean
         get() = accessPolicy == AccessPolicy.ALWAYS_READABLE
     override val includeInShortDump: Boolean = true
+
+    override fun accessInvariantsForPredicate(): TypeInvariantEmbedding =
+        if (symbol.isVal) {
+            FieldAccessTypeInvariantEmbedding(this, PermExp.WildcardPerm())
+        } else {
+            FieldAccessTypeInvariantEmbedding(this, PermExp.FullPerm())
+        }
```
Comment (jesyspa) [inline plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/FieldEmbedding.kt:74]: This refers to the permissions inhaled, right?

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/TypeEmbedding.kt [hunk 3]
```diff
@@ -65,7 +75,8 @@ interface TypeEmbedding {
     fun accessInvariants(): List<TypeInvariantEmbedding> = emptyList()
 
     // Note: this function will replace accessInvariants when nested unfold will be implemented
-    fun predicateAccessInvariant(): TypeInvariantEmbedding? = null
+    // Note: for the moment, unique predicates are not used and so TYPE_PREDICATE can be the default value
```
Comment (jesyspa) [inline plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/TypeEmbedding.kt:78]: Are these sufficiently similar to warrant having the same function?  Is there any case where you may want one or the other?  It feels like all the implementations will start with a branch.

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/FieldEmbedding.kt (hunk 2 above)
Comment (francescoo22) [reply plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/FieldEmbedding.kt:74]: This tells us the way the field is accessed within the body of a predicate.

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/TypeEmbedding.kt (hunk 3 above)
Comment (francescoo22) [reply plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/TypeEmbedding.kt:78]: Do you mean having two functions `TypePredicateAccessInvariant` and `UniquePredicateAccessInvariant`, right?

I'm still not sure about what's the best thing to do here. Probably it will be clearer after adding more functionalities to unique predicates. I think that we can keep this for the moment and decide whether to split it or not in next PR.

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/FieldEmbedding.kt (hunk 2 above)
Comment (jesyspa) [reply plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/FieldEmbedding.kt:74]: I'm sorry, I don't follow.  Could you explain where it shows up?

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/TypeEmbedding.kt [hunk 4]
```diff
@@ -231,47 +236,82 @@ data class ClassTypeEmbedding(val className: ScopedKotlinName, val isInterface:
...
+            (accessFields + accessFieldSharedPredicates + accessFieldUniquePredicates + accessSuperTypesPredicates + fieldsProvenInvariants).toConjunction()
+        return Predicate(uniquePredicateName, listOf(subjectEmbedding.toLocalVarDecl()), body.pureToViper(toBuiltin = true))
+    }
+
     // TODO: incorporate generic parameters.
     override val name = object : MangledName {
         override val mangled: String = "T_class_${className.mangled}"
     }
 
+    private val uniquePredicateName = object : MangledName {
```
Comment (francescoo22) [inline plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/TypeEmbedding.kt:311]: I think we can also mangle the name of the shared predicate in the same way. Since this would modify a lot of tests, probably it's better to do it in a separate PR.

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/TypeEmbedding.kt [hunk 5]
```diff
@@ -65,7 +65,9 @@ interface TypeEmbedding {
     fun accessInvariants(): List<TypeInvariantEmbedding> = emptyList()
 
     // Note: this function will replace accessInvariants when nested unfold will be implemented
-    fun predicateAccessInvariant(): TypeInvariantEmbedding? = null
+    // Note: for the moment, unique predicates are not used and so TYPE_PREDICATE can be the default value
```
Comment (jesyspa) [inline plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/TypeEmbedding.kt:68]: Outdated comment?

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/TypeEmbedding.kt [hunk 6]
```diff
@@ -231,47 +236,82 @@ data class ClassTypeEmbedding(val className: ScopedKotlinName, val isInterface:
...
-            .filterIsInstance<ClassTypeEmbedding>()
+        val accessSuperTypesPredicates = classSuperTypes
             .map { PredicateAccessTypeInvariantEmbedding(it.name, PermExp.WildcardPerm()).fillHole(subjectEmbedding) }
 
         val body = (accessFields + accessFieldPredicates + accessSuperTypesPredicates + fieldsProvenInvariants).toConjunction()
         return Predicate(name, listOf(subjectEmbedding.toLocalVarDecl()), body.pureToViper(toBuiltin = true))
     }
 
+    private fun initUniquePredicate(): Predicate {
+        val filteredFields = fields.values.filterIsInstance<UserFieldEmbedding>()
```
Comment (jesyspa) [inline plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/TypeEmbedding.kt:278]: Why is it not necessary to exclude always-readable fields here?

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/TypeEmbedding.kt [hunk 7]
```diff
@@ -231,47 +236,82 @@ data class ClassTypeEmbedding(val className: ScopedKotlinName, val isInterface:
...
+
+        val accessFieldSharedPredicates = filteredFields
+            .mapNotNull { it.type.sharedPredicateAccessInvariant()?.fillHole(PrimitiveFieldAccess(subjectEmbedding, it)) }
+
+        val accessFieldUniquePredicates = filteredFields
+            .filter { it.isUnique }
+            .mapNotNull { it.type.uniquePredicateAccessInvariant()?.fillHole(PrimitiveFieldAccess(subjectEmbedding, it)) }
+
+        val fieldsProvenInvariants = filteredFields
+            .flatMap { it.type.provenInvariants().fillHoles(PrimitiveFieldAccess(subjectEmbedding, it)) }
```
Comment (jesyspa) [inline plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/TypeEmbedding.kt:296]: Is there a case that this isn't already present in the shared predicate?

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/TypeEmbedding.kt (hunk 6 above)
Comment (francescoo22) [reply plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/TypeEmbedding.kt:278]: Let's consider this example:
```kt
class A(@unique val x: T)
```
Here we need to include `x` because we want to access its unique predicate
```
UniqueA(this: Ref) { acc(this.x, wildcard) && UniqueT(this.x) }
```

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/TypeEmbedding.kt (hunk 7 above)
Comment (francescoo22) [reply plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/TypeEmbedding.kt:296]: `var` fields are not included in the shared predicates.
`provenInvariants` of `val` fields appear both in the shared and the unique predicate, but it shouldn't be a problem.
Do you think it can be worth to remove `provenInvariants` of `val` fields from the unique predicate?

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/TypeEmbedding.kt (hunk 6 above)
Comment (francescoo22) [reply plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/TypeEmbedding.kt:277]: yes, probably using a predicate builder is a good idea. I'll try to implement it.

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/TypeEmbedding.kt [hunk 8]
```diff
@@ -231,47 +235,122 @@ data class ClassTypeEmbedding(val className: ScopedKotlinName, val isInterface:
...
+            body.addAll(fields.flatMap(transform))
+        }
+    }
+
+    private fun initSharedPredicate() = ClassPredicateBuilder(this).build(name) {
+        filterFields { it.accessPolicy == AccessPolicy.ALWAYS_READABLE }
+        accessFields()
+        accessFieldsSharedPredicate()
+        accessSuperTypesPredicates(PermExp.WildcardPerm()) { name }
+        fieldsProvenInvariants()
```
Comment (jesyspa) [inline plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/TypeEmbedding.kt:334]: To be honest, I don't really like this design.  The benefit of a builder is that you should be able to specify what the predicate should contain in an easy-to-read fashion.  Here, you still need to know what all these functions do; the builder doesn't really add anything.

I think one problem we're having is that our naming isn't great: `proven`, `access`, it is all a bit vague.  Maybe it would be good to take a step back, identify the permissions we have, and then write a builder that lets you select fields/supertypes and then add the relevant permissions for them?

**Review by jesyspa — COMMENTED** [review]
Mostly stylistic comments, the general structure looks good, but I'm not yet convinced this is what we should be putting in the unique/shared predicates.
Let's have a brainstorming session for where and how to include subtype invariants and such?

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/ClassPredicateBuilder.kt [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    fun onType(action: TypeInvariantsBuilder.() -> Unit) {
+        val builder = TypeInvariantsBuilder(field.type)
+        builder.action()
+        assertions.addAll(builder.toInvariantsList().fillHoles(PrimitiveFieldAccess(subject, field)))
+    }
+
+    fun access() {
+        val perm = if (field.symbol.isVal) PermExp.WildcardPerm() else PermExp.FullPerm()
```
Comment (jesyspa) [inline plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/ClassPredicateBuilder.kt:71]: By the way, I would move this choice to user code (i.e. make it a parameter of `access`).

**Review by jesyspa — APPROVED** [review]
Discussed via call, some TODOs but LGTM overall
